jlm/hls/backend/rvsdg2rhls/UnusedStateRemoval.cpp

```diff
@@ -129,53 +129,48 @@ RemovePassthroughArgument(const rvsdg::RegionArgument & argument)
   region->node()->RemoveOutput(outputIndex);
 }
 
-// If this output has a single user and that single user happens to be
-// the exit variable of this gamma node, then return it.
-static std::optional<rvsdg::GammaNode::ExitVar>
-TryGetSingleUserExitVar(rvsdg::GammaNode & gammaNode, rvsdg::output & argument)
+static void
+RemoveUnusedStatesFromGammaNode(rvsdg::GammaNode & gammaNode)
 {
-  if (argument.nusers() == 1)
+  auto entryvars = gammaNode.GetEntryVars();
+  for (int i = entryvars.size() - 1; i >= 0; --i)
   {
-    rvsdg::input * user = *argument.begin();
-    if (rvsdg::TryGetRegionParentNode<rvsdg::GammaNode>(*user) == &gammaNode)
+    size_t resultIndex = 0;
+    auto argument = entryvars[i].branchArgument[0];
+    if (argument->nusers() == 1)
     {
-      return gammaNode.MapBranchResultExitVar(*user);
+      auto result = dynamic_cast<rvsdg::RegionResult *>(*argument->begin());
+      resultIndex = result ? result->index() : resultIndex;
     }
-  }
-  return std::nullopt;
-}
-
-static void
-RemoveUnusedStatesFromGammaNode(rvsdg::GammaNode & gammaNode)
-{
-  std::vector<rvsdg::GammaNode::EntryVar> deadEntryVars;
-  std::vector<rvsdg::GammaNode::ExitVar> deadExitVars;
 
-  for (const auto & entryvar : gammaNode.GetEntryVars())
-  {
-    std::optional<rvsdg::GammaNode::ExitVar> exitvar0 =
-        TryGetSingleUserExitVar(gammaNode, *entryvar.branchArgument[0]);
-
-    bool shouldRemove = exitvar0
-                     && std::all_of(
-                            entryvar.branchArgument.begin(),
-                            entryvar.branchArgument.end(),
-                            [&gammaNode, &exitvar0](rvsdg::output * argument) -> bool
-                            {
-                              auto exitvar = TryGetSingleUserExitVar(gammaNode, *argument);
-                              return exitvar && exitvar->output == exitvar0->output;
-                            });
+    bool shouldRemove = true;
+    for (size_t n = 0; n < gammaNode.nsubregions(); n++)
+    {
+      auto subregion = gammaNode.subregion(n);
+      shouldRemove &=
+          IsPassthroughArgument(*subregion->argument(i))
```

Comment:
Should `entryvars.branchArguments[n]` be used instead of looking up the argument from the subregion?